### PR TITLE
arch: arc: implement mpu gap filling

### LINKED ISSUE #19067: non-overlapping MPU gap-filling needs to be optional
ARMv8, and ARC MPU v3 are newer MPUs which allow for great flexibility on MPU region sizes, but regions are not allowed to overlap.

Currently, MPU regions are set up to control supervisor mode access, so that supervisor mode reads/writes to invalid memory generate faults. Implementing overlapping regions for memory domain requires that these base regions be split.

Unfortunately, this gap-filling has drawbacks:
1. The number of MPU regions available to applications is cut roughly in half, due to the need for additional regions to fill in the gaps
2. The gap-filling introduces overhead, currently during context switch

The overhead can be looked at (see #15223) but halving the number of free regions may in a lot of cases be a dealbreaker for applications.

This gap-filling is NOT needed to properly support user mode. This is a feature to catch stray memory access in supervisor mode and is orthogonal to CONFIG_USERSPACE requirements.

Add a Kconfig option for these MPUs which allows this to be disabled, maximizing the number of memory domain partitions available to applications.

### PR DESCRIPTION
**Pull request by vonhust**
This PR implements MPU gap filling for ARC, ARM alread did it

For the discussions on mpu gap filling, pls refer comments in  #19067, #15223 and #20152.

ARC related work in  #19067 

and improves #15223

### CODE REVIEW

**Review by abrodkin — CHANGES_REQUESTED** [review]
To that looks like a pretty significant change given its diff size so I'd like to see explanation of what we do here and why.

1. What is that `MPU_GAP_FILLING`? For me it makes no sense and I guess other readers will have the same question. Is that something related to ARC HW, or purely Zephyr something?
2. What are those gaps, why do they appear and why do we need/want to do something with them?
3. Is that just an enhancement or a fix or entirely new feature?
4. What is high-level explanation of your implementation? W/o that it's hard to judge your approach unless reviewer spends the same time as you did look at all that and trying to implement the same thing him/her-self. I guess nobody is going to do that, right? Then what else other than coding style might be discussed in this review? :)

Please refer to this commit as a pretty good example of a commit message explaining both our problem at hand and our strategy to resolve it: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=ec9b4feb1e41587c15d43d237844193318389dc3

**Comment by vonhust** [comment]
@abrodkin  there are already a lot of discussions on mpu gap filling, pls refer #19067, #15223 and #20152.

**Comment by abrodkin** [comment]
@vonhust that's good you have those references but:
1. This particular patch doesn't mention either of them. So how reader of that patch (when it is already accepted) could figure out there were relevant discussions somewhere? So maybe add references to those isseus in the commit message?
2. Even if there're references it's good to provide compact summary of it right in the commit message as it really simplifies review. Also what if Zephyr moves from GitHub to GitLab or Gerrit and all your references to GitHub issues become broken links!

File: tests/kernel/mem_protect/userspace/testcase.yaml [hunk 1]
```diff
@@ -3,8 +3,8 @@ tests:
     filter: CONFIG_ARCH_HAS_USERSPACE
     tags: kernel security userspace ignore_faults
     min_ram: 36
-  kernel.memory_protection.userspace.armv8m_gap_filling:
-    arch_whitelist: arm
+  kernel.memory_protection.userspace.gap_filling:
+    arch_whitelist: arm arc
```
Comment (andrewboie) [inline tests/kernel/mem_protect/userspace/testcase.yaml]: do we even need this whitelist? seems like CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS filter ought to be enough now

Comment (vonhust) [reply tests/kernel/mem_protect/userspace/testcase.yaml]: already changed.
it makes sense to make it more generic to covery any  archs which support CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS.

**Comment by dleach02** [comment]
sorry, went to comment then cancel comment and clicked close instead.

**Review by ioannisg — COMMENTED** [review]
Didn't look inside the ARC application but the rest seems ok to me.

**Comment by vonhust** [comment]
@abrodkin @ruuddw we need to get this PR merged for release 2.2,  so pls approve if no other comments

**Review by abrodkin — COMMENTED** [review]
Even though there're discussions of MPU gap filling in some GitHub issues mentioned in this PR (as I already mentioned Zephyr might move from GitHub to some other place and all the references will be lost for good) I'd like to have explanation of our design either in a separate .rst document or even better right here in code.

That's because we're getting more and more features, code becomes very complex and few months down the line we won't be able to understand what do we do right here and why. This is especially true if another engineer will be looking at this code.

@vonhust do you think we may get it done and still squeeze in 2.2 release?

**Comment by vonhust** [comment]
> Even though there're discussions of MPU gap filling in some GitHub issues mentioned in this PR (as I already mentioned Zephyr might move from GitHub to some other place and all the references will be lost for good) I'd like to have explanation of our design either in a separate .rst document or even better right here in code.
> 
> That's because we're getting more and more features, code becomes very complex and few months down the line we won't be able to understand what do we do right here and why. This is especially true if another engineer will be looking at this code.
> 
> @vonhust do you think we may get it done and still squeeze in 2.2 release?

@abrodkin , I understand your concern. It's nor trivial to have a quick understanding of code without deatiled comments.  If possible, @andrewboie @ioannisg  could we work together to write a generice doc about mpu driver implementation?  It may take some time.

In short time, I will try to add more comments in code .  We should get this in 2.2 release

**Comment by andrewboie** [comment]
FWIW I don't see gap-filling as particularly urgent, it's orthogonal to user mode

**Review by ruuddw — APPROVED** [review]
The code structure and #ifdef's still seem to be (too) complex and not well documented, but the logic now looks ok to me.

**Review by abrodkin — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

**Comment by ioannisg** [comment]
> In short time, I will try to add more comments in code . We should get this in 2.2 release

@vonhust  Adding TSC label, to see if we can get this to 2.2; probably need to discuss this on this week's TSC.

**Pull request merged**